### DISCUSSION TITLE
Autogeneration of files using custom_target

### DISCUSSION THREAD

**Discussion by lkammath** [mesonbuild/meson#9626] [Q&A]
I am trying to generate some header/c files using a custom Perl script using curtom_target

here is the code snippet

```
env = environment()
env.set('SRCPATH', project_source_root, 'src')
auto_generate_src = run_command('env', env: env)

message('Autogenerating of ecmd & client CAPI code')
makedll_cmd = find_program('tool/mkscripts/autogen.pl')
foreach code_gen : ['A.C', 'A.H', 'X.H']
   autogen_target = custom_target(
     code_gen,
     output: code_gen,
     input:'autogen.pl',
     command: [perl, '@INPUT@','--opt','@OUTPUT@'])
endforeach
```

The autogen script creates the files in the path where SRCPATH environment pointing to . that is how the perl script is written

I am unable make the meson snippet work for some reason. Could some one throw some light here ?

**Comment by eli-schwartz** [comment]
Since 0.57.0, custom_target() supports the `env` kwarg as well.

Running the /usr/bin/env program using run_command() at configure time using a modified env did nothing -- Unix doesn't allow you to define environment variables in a parent process by forking and executing a command, and even if it did, the /usr/bin/env program would just print a list of currently known environment variables to stdout.

Before meson 0.57.0 you could have done:

```
srcpath = 'SRCPATH=@0@:@1@'.format(project_source_root, 'src')
foreach code_gen : ['A.C', 'A.H', 'X.H']
   autogen_target = custom_target(
     code_gen,
     output: code_gen,
     input:'autogen.pl',
     command: ['env', srcpath, perl, '@INPUT@','--opt','@OUTPUT@'])
endforeach
```

That would make the custom_target run the env command, which would fork and modify the environment to set SRCPATH, and, since a command is provided, run that command instead of printing all known variables by default. As your overall objective is to run perl with a modified env ***at build time***, this would work.

(But I don't understand the purpose of setting `SRCPATH=/path/to/project/root:src`, did you mean to use `env.set(..., separator: '/')` by any chance?)

...

As a matter of curiosity:

> that is how the perl script is written

Why not fix the perl script? Scripts can be fixed. It doesn't take the ouptut file as an environment variable `OUTPUT=A.C`, so why does it take `SRCPATH` as an environment variable? Using env support in meson means meson needs to run a python wrapper script in your build.ninja, which first sets up the environment variables (in a cross platform way) and then runs autogen.pl. Much simpler to do all the work in autogen.pl itself.
